src/internals/ir/callGraph.ts

```diff
@@ -0,0 +1,169 @@
+import { unreachable } from "../util";
+import { TactASTStore } from "./astStore";
+import { IdxGenerator } from "./indices";
+import { forEachExpression } from "../tact/iterators";
+import {
+  AstFunctionDef,
+  AstReceiver,
+  AstContractInit,
+  AstExpression,
+  AstStaticCall,
+  AstMethodCall,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+export type CGNodeId = number;
+export type CGEdgeId = number;
+
+export class CGEdge {
+  public idx: CGEdgeId;
+  constructor(
+    public src: CGNodeId,
+    public dst: CGNodeId,
+  ) {
+    this.idx = IdxGenerator.next("cg_edge");
+  }
+}
+
+export class CGNode {
+  public idx: CGNodeId;
+  public inEdges: Set<CGEdgeId> = new Set();
+  public outEdges: Set<CGEdgeId> = new Set();
+  constructor(
+    public astId: number | undefined,
+    public name: string,
+  ) {
+    this.idx = IdxGenerator.next("cg_node");
+  }
+}
+
+export class CallGraph {
+  private nodeMap: Map<CGNodeId, CGNode> = new Map();
+  private edgesMap: Map<CGEdgeId, CGEdge> = new Map();
+  private nameToNodeId: Map<string, CGNodeId> = new Map();
+
+  public getNodes(): Map<CGNodeId, CGNode> {
+    return this.nodeMap;
+  }
+
+  public getEdges(): Map<CGEdgeId, CGEdge> {
+    return this.edgesMap;
+  }
+
+  build(astStore: TactASTStore): CallGraph {
+    this.addFunctionsToNodes(astStore);
+    this.analyzeFunctionCalls(astStore);
+    return this;
+  }
+
+  private addFunctionsToNodes(astStore: TactASTStore) {
+    for (const func of astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        const node = new CGNode(func.id, funcName);
+        this.nodeMap.set(node.idx, node);
+        this.nameToNodeId.set(funcName, node.idx);
+      }
+    }
+  }
+
+  private analyzeFunctionCalls(astStore: TactASTStore) {
+    for (const func of astStore.getFunctions()) {
+      const callerId = this.getNodeIdByName(this.getFunctionName(func));
+      if (callerId !== undefined) {
+        this.processFunctionBody(func, callerId);
+      }
+    }
+  }
+
+  private getFunctionName(
+    func: AstFunctionDef | AstReceiver | AstContractInit,
+  ): string {
+    switch (func.kind) {
+      case "function_def":
+        return func.name.text;
+      case "contract_init":
+        return `contract_init_${func.id}`;
+      case "receiver":
+        return `receiver_${func.id}`;
+      default:
+        unreachable(func);
+    }
+  }
+
+  private getNodeIdByName(name: string): CGNodeId | undefined {
+    return this.nameToNodeId.get(name);
+  }
+
+  private processFunctionBody(
```

Comment:
I would not create a separate function for this. It is used once and contains only one statement.